Track and save interval histograms per agent

lampajr opened the pull request:
<!-- If your PR fixes an open issue, use `Closes #435` or `Fixes #435` to link your PR with the issue. #435 stands for the issue number you are fixing -->

This is still a draft proposal on how we could save HDR histograms per interval and agent.

## Fixes Issue

Fixes https://github.com/Hyperfoil/Hyperfoil/issues/588

## Changes proposed

Tracks interval histograms so that they can be retrieved for further analysis, e.g., using https://hdrhistogram.github.io/HdrHistogramJSDemo/logparser.html or https://hdrhistogram.github.io/HdrHistogram/plotFiles.html

- [x] the tracking is disabled by default but can be enabled via global controller config `-Dio.hyperfoil.controller.persist.hdr=true` 
- [x] tracking interval histograms per agent configured
- [x] tracking global histograms per phase

## Check List (check all the applicable boxes) <!-- Follow the above conventions to check the box -->

- [x] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.

> Note that the documentation is located at [github.com/Hyperfoil/hyperfoil.github.io](https://github.com/Hyperfoil/hyperfoil.github.io/) 

<details>
<summary>
How to backport a pull request to the current <em>stable</em> branch?
</summary>

In order to automatically create a **backporting pull request** please add `backport` label to the current pull request.

Then, as soon as the pull request is merged into the `master` branch, the process will try to automatically open a backporting pull request against the _stable_ branch. If something goes wrong, a comment will be added in the original pull request such that all people involved get notified.

> The `backport` label can be also added after the pull  request has been merged.

> If the process fails due to temporary problems, such as connectivity problems, consider removing and re-adding the `backport` label.
</details>

lampajr committed:
Track and save interval histograms per agent

Signed-off-by: Andrea Lamparelli <a.lamparelli95@gmail.com>

lampajr commented:
This is an example of how it could look like when uploaded to https://hdrhistogram.github.io/HdrHistogramJSDemo/logparser.html

<img width="1105" height="836" alt="image" src="https://github.com/user-attachments/assets/1fccb917-4bd4-422b-ac34-a8ad1d793426" />

There is still something not completely clear that I need to double check

lampajr commented on clustering/src/main/java/io/hyperfoil/controller/Data.java:
```diff
@@ -31,6 +32,8 @@ final class Data {
    final Map<String, StatisticsSnapshot> perAgent = new HashMap<>();
    final Map<String, IntObjectMap<StatisticsSnapshot>> lastStats = new HashMap<>();
    final List<StatisticsSummary> series = new ArrayList<>();
+   final Map<String, List<Histogram>> intervalHistogramsPerAgent = new HashMap<>();
+   final List<Histogram> intervalHistograms = new ArrayList<>();
```
Had to track the histograms at `Data` level, still nor sure whether this could introduce problems 🤔

franz1981 replied:
2 things:
- it would be great to not collect anything unless required i.e. if the hdr output is configured
- it would be good to pre-size the list based on the duration

The last is a minor TBH since I won't expect total duration to exceed 15 minutes or so, which means 15 * 60 histograms.

franz1981 commented:
On Monday I will take a look debugging it too 🙏 well done 👍

lampajr commented:
> I think you have to manually adjust the intervals of the first non warmup histogram (that start) to link with the last one of non warm up?

That's what I was thinking but I tried to debug to it and the start/end timestamp look already good.. I will double check on Monday but for sure there is something wrong in some start time otherwise I don't get why both tags show the same start time

franz1981 commented on clustering/src/main/java/io/hyperfoil/controller/HdrWriter.java:
```diff
@@ -0,0 +1,63 @@
+   }
+
+   private static String getTag(String agent, Data d) {
+      return agent + "/" + d.phase + "/" + d.metric + "/" + (d.isWarmup ? "warmup" : "");
```
this is nice, good idea

franz1981 reviewed (commented):
I think it's good - just need to fix the ranges, and crate some config option to not enable this by default, but overall looks good.

stalep commented on clustering/src/main/java/io/hyperfoil/controller/Data.java:
```diff
@@ -27,26 +28,30 @@ final class Data {
       this.metric = metric;
       this.windowSlas = periodSlas;
       this.totalSlas = totalSlas;
+      this.trackIntervalHistograms = trackIntervalHistograms;
```
With regards to Franz's comment on not allocating if we're not saving we can add a check here if `trackIntervalHistograms == true` and instantiate `intervalHistogramsPerAgent` and `intervalHistograms` only if it is?

lampajr replied:
yeah that's something I can add as well, at the moment they are always instantiated but kept empty if `trackIntervalHistograms == true`

stalep committed:
Address review feedback: use primitive boolean and lazy-allocate histogram collections

- Change Boolean to boolean for trackIntervalHistograms throughout
- Only allocate intervalHistogramsPerAgent and intervalHistograms when
  tracking is enabled, using Collections.emptyMap()/emptyList() otherwise
- Fix double semicolon in Data field declaration

stalep commented:
## PR updated

Rebased onto current `master` (resolved conflicts in `ControllerVerticle` and `BaseScenarioTest`) and addressed review feedback:

- **`Boolean` → `boolean`** for `trackIntervalHistograms` in `Controller`, `StatisticsStore`, `JsonLoader`, and `Data`
- **Lazy allocation** — `intervalHistogramsPerAgent` and `intervalHistograms` are only allocated when `trackIntervalHistograms == true`; otherwise they use `Collections.emptyMap()` / `Collections.emptyList()`, so there is zero overhead when HDR persistence is disabled
- **Fixed double semicolon** in `Data` field declaration

Still open for follow-up (can be separate issues):
- REST API endpoint to download `.hgrm` files (the issue title mentions "display/download")
- Consider making this a per-benchmark YAML config rather than a global controller system property
- More thorough test coverage for `HdrWriter` with actual recorded statistics

stalep commented:
@diegolovison or @franz1981, if you have time to have a look at this again?
